flair/models/biomedical_entity_linking.py

```diff
@@ -0,0 +1,988 @@
+
+    def embed_dense(
+        self,
+        names: np.ndarray,
+        show_progress: bool = False,
+        batch_size: int = 2048,
+    ) -> np.ndarray:
+        """
+        Embedding data into dense representations for SapBert
+        :param names: np.array of names
+        :param show_progress: bool to toggle progress bar
+        :param batch_size: batch size
+        :return dense_embeds: list of dense embeddings of the names
+        """
+        self.encoder.eval()  # prevent dropout
+
+        dense_embeds = []
+
+        with torch.no_grad():
+            if show_progress:
+                iterations = tqdm(
+                    range(0, len(names), batch_size),
+                    desc="Calculating dense embeddings for dictionary",
+                )
+            else:
+                iterations = range(0, len(names), batch_size)
+
+            for start in iterations:
+
+                # make batch
+                end = min(start + batch_size, len(names))
+                batch = [Sentence(name) for name in names[start:end]]
+
+                # embed batch
+                self.encoder.embed(batch)
+
+                dense_embeds += [
+                    name.embedding.cpu().detach().numpy() for name in batch
+                ]
+
+                if flair.device.type == "cuda":
+                    torch.cuda.empty_cache()
+
+        return np.array(dense_embeds)
+
+    def get_sparse_similarity_scores(
+        self,
+        query_embeds: np.ndarray,
+        dict_embeds: np.ndarray,
+        cosine: bool = False,
+        normalise: bool = False,
+    ) -> np.ndarray:
+        """
+        Return score matrix
+        :param query_embeds: 2d numpy array of query embeddings
+        :param dict_embeds: 2d numpy array of query embeddings
+        :param score_matrix: 2d numpy array of scores
+        """
+        if cosine:
+            score_matrix = cosine_similarity(query_embeds, dict_embeds)
+        else:
+            score_matrix = np.matmul(query_embeds, dict_embeds.T)
+
+        if normalise:
+            score_matrix = (score_matrix - score_matrix.min()) / (
+                score_matrix.max() - score_matrix.min()
+            )
+
+        return score_matrix
+
+    def retrieve_sparse_topk_candidates(
+        self,
+        score_matrix: np.ndarray,
+        topk: int,
+        batch_size: int = 128,
+        show_progress: bool = False,
+    ) -> Tuple[np.ndarray, np.ndarray]:
+        """
+        Return sorted topk idxes (descending order)
+        :param score_matrix: 2d numpy array of scores
+        :param topk: number of candidates to retrieve
+        :return res: d numpy array of ids [# of query , # of dict]
+        :return scores: numpy array of top scores
+        """
+
+        res = None
+        scores = None
+        for i in tqdm(
+            np.arange(0, score_matrix.shape[0], batch_size), disable=not show_progress
+        ):
+            score_matrix_tmp = torch.tensor(
```

Comment:
Why go to torch if then you go back to numpy?

Reply:
There's a vectorized implementation [here](https://github.com/dmis-lab/BioSyn/blob/master/src/biosyn/biosyn.py#L157)